### ISSUE TITLE
zstd fails on message consumption: UnsupportedCompressionType

### ISSUE THREAD

**Issue by gertvdijk** [fede1024/rust-rdkafka#201]
I've set up rdkafka 0.22.0 in my project with Cargo.toml like this, to use the [recently added zstd compression](https://github.com/fede1024/rust-rdkafka/issues/136) feature:

```toml
[dependencies]
rdkafka = { version = "0.22.0", features = ["cmake-build", "zstd"] }
```
It seems to compile fine on Debian Stretch with zstd from backports, `libzstd-dev:amd64` `1.3.8+dfsg-3~bpo9+1`.

Connecting to Kafka Cluster version 2.3.0 with brokers set to `compression.type=zstd` in `server.properties` and consuming messages fails with the following exception:

```
Message consumption error: UnsupportedCompressionType (Unsupported compression type))
```

The binary built and used *does* have zstd compiled in statically:
```
$ strings /path/to/binary | grep -Fi zstd
```
shows many many lines of the ZSTD function signatures.

I wonder what I could do to debug this further. Does anyone have a successful zstd enabled setup?

The other processes connecting to the Kafka cluster have no issue using zstd (JDK).

**Comment by benesch** [comment]
Hmm, AFAICT zstd compression is working fine on the latest master. If I create some records via

```
kafka-console-producer --broker-list localhost:9092 --topic compress --compression-codec zstd
```

I can read them out with:

```
cargo run --features cmake-build,zstd --example simple_consumer -- --topics compress
```

**Comment by benesch** [comment]
To be honest I'm not sure what component produces the unsupported compression type error message. It's not rust-rdkafka, and it doesn't seem to be librdkafka either. I guess it's the Kafka broker itself—but I'm not sure what librdkafka is sending along that causes the Kafka broker to think the compression type is unsupported. Strange.

**Comment by benesch** [comment]
Yeah the conditions under which this unsupported compression type error is generated seem very strange: https://github.com/apache/kafka/blob/b94c7f479b917d4ec602c31a24f11390627c479b/core/src/main/scala/kafka/server/KafkaApis.scala#L648-L700

To debug this further, you could look for either of those `trace` messages in the snippet above in your Kafka logs. I'm also curious what your consumer code looks like. Are you setting unusual configuration parameters, perhaps?

**Comment by gertvdijk** [comment]
I've noticed that another Rust-rdkafka component seems to produce & consume just fine. The difference for those is that messages on those topics are only produced by Rust-rdkafka as well, whereas the problem appears from topics where messages are produced from components using Java/JDK. :frowning: 

The Rust code looks like this. I don't think the options used are very relevant, though.

```
let consumer = ClientConfig::new()
        .set("group.id", group_id)
        .set("bootstrap.servers", brokers)
        .set("message.timeout.ms", "5000")
        .set("request.required.acks", "all")
        .set("max.in.flight.requests.per.connection", "1")
        .set("batch.num.messages", "50000")
        .set("queue.buffering.max.messages", "1000000")
        .set("queue.buffering.max.ms", "100")
        .set("statistics.interval.ms", "1000")
        .set("enable.partition.eof", "false")
        .set("enable.auto.commit", "false")
        .create_with_context::<_, BaseConsumer<StatisticsContext>>(context_in.clone())
        .expect("Consumer creation failed");
[...]
    let (_low, initial_high_watermark) = consumer
        .fetch_watermarks(topic_in, TOPIC_PARTITION_ID, Duration::from_secs(1))
        .expect("failed to fetch watermarks.");

    let mut mapping = HashMap::new();
[...]
    mapping.insert((topic_in.to_owned(), TOPIC_PARTITION_ID), Offset::Beginning);
    let tpl = TopicPartitionList::from_topic_map(&mapping);
    consumer.assign(&tpl)?;
[...]
   let mut read_topic_complete = initial_high_watermark == 0;
[...]
   if let Some(result) = consumer.poll(Duration::from_millis(100)) {
      let message = result?;
[...]
      let payload_bytes = message
                .payload()
                .ok_or_else(|| MyError::KafkaMessageValueNotSet(KafkaMessageRef::from_message(&message)))?;
```

**Comment by benesch** [comment]
Did you have any luck tracking this down?

**Comment by gertvdijk** [comment]
@benesch No, I haven't...

<sup>FWIW, In the end I found out that compression wasn't beneficial at all for what my/our situation to begin with. My messages are already very small being protobuf encoded and the similarity is between the messages, not in the context of a single one. This makes virtual filesystem level compression way more beneficial for my goal slimming down the disk usage footprint.</sup>

**Comment by athre0z** [comment]
Encountered a similar problem today where I randomly started to get crashes with:

```
Message consumption error: BadCompression (Local: Invalid compressed data)
```

when using zstd compression. It was reproducible for a while, but after clearing the Kafka topic a few times, trying to debug it, it then it magically stopped crashing despite a perfectly empty git diff.

Producer config:
```rust
pub fn create_producer(brokers: &str) -> Result<FutureProducer> {
    ClientConfig::new()
        .set("bootstrap.servers", brokers)
        .set("message.timeout.ms", "3000")
        .set("compression.codec", "zstd")
        .set("compression.level", "8")
        .set("queue.buffering.max.ms", "1000")
        .set("queue.buffering.max.messages", "1000000")
        .set("message.max.bytes", "16777216")
        .create()
        .map_err(Into::into)
}
```

Consumer:

```rust
let consumer: &StreamConsumer<_> = &ClientConfig::new()
        .set("group.id", &sink_topic)
        .set("bootstrap.servers", brokers)
        .set("enable.partition.eof", "false")
        .set("session.timeout.ms", "6000")
        .set("enable.auto.commit", "false")
        .set("max.partition.fetch.bytes", "16777216")
        .create()?;
```

**Comment by benesch** [comment]
@athre0z that's a known bug in zstd v1.4.5. Try downgrading to zstd v1.4.4. The bug will be fixed in zstd v1.4.6, which hasn't shipped yet, but you could build zstd from source in the meantime.

See:
  * https://github.com/edenhill/librdkafka/issues/2672
  * https://github.com/facebook/zstd/issues/2222

Pretty sure the two issues are unrelated though, as zstd v1.4.5 is more recent than this issue.

**Comment by athre0z** [comment]
Thanks for the quick response, that's good to know!

**Comment by benesch** [comment]
Since we don't know how to reproduce the original issue, I'm going to consider this solved once #316 lands, which bumps to a version of zstd-sys that does not exhibit this bug.
